File: HPWorkpath_v1.6.3/Samples/ExampleAPIServices/source/CopySample/src/com/hp/workpath/sample/copysample/fragments/StampOptionDialogPreference.java
package com.hp.workpath.sample.copysample.fragments;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Handler;
import android.util.AttributeSet;

import androidx.preference.DialogPreference;
import androidx.preference.PreferenceManager;

import com.google.gson.Gson;
import com.hp.workpath.api.copier.CopyAttributes;
import com.hp.workpath.api.copier.StampOption;
import com.hp.workpath.api.copier.StampPolicyType;
import com.hp.workpath.api.copier.StampType;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import static com.hp.workpath.sample.copysample.fragments.CopyConfigureFragment.PREF_STAMP;

public class StampOptionDialogPreference extends DialogPreference {

    private List<CopyAttributes.StampPosition> mStampPositionList;
    private Map<CopyAttributes.StampPosition, List<StampType>> mStampTypeListMap;
    private Map<CopyAttributes.StampPosition, List<StampPolicyType>> mStampPolicyTypeListMap;
    private Map<CopyAttributes.StampPosition, List<String>> mStampFontListMap;
    private Map<CopyAttributes.StampPosition, List<Integer>> mStampTextSizeListMap;
    private Map<CopyAttributes.StampPosition, List<String>> mStampTextColorListMap;
    private Map<CopyAttributes.StampPosition, List<Boolean>> mStampWhiteBackGroundListMap;

    private Map<CopyAttributes.StampPosition, StampOption> mStampOptionMap;


    public StampOptionDialogPreference(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public List<CopyAttributes.StampPosition> getmStampPositionList() {
        return mStampPositionList;
    }

    public void setmStampPositionList(List<CopyAttributes.StampPosition> mStampPositionList) {
        this.mStampPositionList = mStampPositionList;
    }

    public Map<CopyAttributes.StampPosition, List<StampType>> getmStampTypeListMap() {
        return mStampTypeListMap;
    }

    public void setmStampTypeListMap(Map<CopyAttributes.StampPosition, List<StampType>> mStampTypeListMap) {
        this.mStampTypeListMap = mStampTypeListMap;
    }

    public Map<CopyAttributes.StampPosition, List<StampPolicyType>> getmStampPolicyTypeListMap() {
        return mStampPolicyTypeListMap;
    }

    public void setmStampPolicyTypeListMap(Map<CopyAttributes.StampPosition, List<StampPolicyType>> mStampPolicyTypeListMap) {
        this.mStampPolicyTypeListMap = mStampPolicyTypeListMap;
    }

    public Map<CopyAttributes.StampPosition, List<String>> getmStampFontListMap() {
        return mStampFontListMap;
    }

    public void setmStampFontListMap(Map<CopyAttributes.StampPosition, List<String>> mStampFontListMap) {
        this.mStampFontListMap = mStampFontListMap;
    }

    public Map<CopyAttributes.StampPosition, List<Integer>> getmStampTextSizeListMap() {
        return mStampTextSizeListMap;
    }

    public void setmStampTextSizeListMap(Map<CopyAttributes.StampPosition, List<Integer>> mStampTextSizeListMap) {
        this.mStampTextSizeListMap = mStampTextSizeListMap;
    }

    public Map<CopyAttributes.StampPosition, List<String>> getmStampTextColorListMap() {
        return mStampTextColorListMap;
    }

    public void setmStampTextColorListMap(Map<CopyAttributes.StampPosition, List<String>> mStampTextColorListMap) {
        this.mStampTextColorListMap = mStampTextColorListMap;
    }

    public Map<CopyAttributes.StampPosition, List<Boolean>> getmStampWhiteBackGroundListMap() {
        return mStampWhiteBackGroundListMap;
    }

    public void setmStampWhiteBackGroundListMap(Map<CopyAttributes.StampPosition, List<Boolean>> mStampWhiteBackGroundListMap) {
        this.mStampWhiteBackGroundListMap = mStampWhiteBackGroundListMap;
    }

    public Map<CopyAttributes.StampPosition, StampOption> getmStampOptionMap() {
        return mStampOptionMap;
    }

    public void setmStampOptionMap(Map<CopyAttributes.StampPosition, StampOption> mStampOptionMap) {
        this.mStampOptionMap = mStampOptionMap;
    }


    public void saveStampOptionMap() {
        SharedPreferences sharedPref = PreferenceManager.getDefaultSharedPreferences(getContext());
        SharedPreferences.Editor editor = sharedPref.edit();
        Gson gson = new Gson();


        for(CopyAttributes.StampPosition stampPosition : mStampOptionMap.keySet()){
            String jsonStampOption = gson.toJson(mStampOptionMap.get(stampPosition));
            editor.putString(PREF_STAMP+stampPosition.name(),jsonStampOption);
        }

        editor.apply();
        setStampSummary();
    }
    private void setStampSummary() {
        List<String> stamptype = new ArrayList<String>();
        SharedPreferences sharedPref = PreferenceManager.getDefaultSharedPreferences(this.getContext());
        Gson gson = new Gson();
        String top_left = sharedPref.getString("pref_stamp" + CopyAttributes.StampPosition.TOP_LEFT, (String) null);
        String top_right = sharedPref.getString("pref_stamp" + CopyAttributes.StampPosition.TOP_RIGHT, (String) null);
        String top_center = sharedPref.getString("pref_stamp" + CopyAttributes.StampPosition.TOP_CENTER, (String) null);
        String bottom_left = sharedPref.getString("pref_stamp" + CopyAttributes.StampPosition.BOTTOM_LEFT, (String) null);
        String bottom_right = sharedPref.getString("pref_stamp" + CopyAttributes.StampPosition.BOTTOM_RIGHT, (String) null);
        String bottom_center = sharedPref.getString("pref_stamp" + CopyAttributes.StampPosition.BOTTOM_CENTER, (String) null);
        StampOption top_left_stampOption = (StampOption) gson.fromJson(top_left, StampOption.class);
        StampOption top_right_stampOption = (StampOption) gson.fromJson(top_right, StampOption.class);
        StampOption top_center_stampOption = (StampOption) gson.fromJson(top_center, StampOption.class);
        StampOption bottom_left_stampOption = (StampOption) gson.fromJson(bottom_left, StampOption.class);
        StampOption bottom_right_stampOption = (StampOption) gson.fromJson(bottom_right, StampOption.class);
        StampOption bottom_center_stampOption = (StampOption) gson.fromJson(bottom_center, StampOption.class);

        if (top_left_stampOption != null) {
            if (!top_left_stampOption.type.name().equals("NONE")) {
                stamptype.add(CopyAttributes.StampPosition.TOP_LEFT.name());
            } else {
                stamptype.remove(CopyAttributes.StampPosition.TOP_LEFT.name());
            }
        }
        if (top_right_stampOption != null) {
            if (!top_right_stampOption.type.name().equals("NONE")) {
                stamptype.add(CopyAttributes.StampPosition.TOP_RIGHT.name());
            } else {
                stamptype.remove(CopyAttributes.StampPosition.TOP_RIGHT.name());
            }
        }
        if (top_center_stampOption != null) {
            if (!top_center_stampOption.type.name().equals("NONE")) {
                stamptype.add(CopyAttributes.StampPosition.TOP_CENTER.name());
            } else {
                stamptype.remove(CopyAttributes.StampPosition.TOP_CENTER.name());
            }
        }
        if (bottom_left_stampOption != null) {
            if (!bottom_left_stampOption.type.name().equals("NONE")) {
                stamptype.add(CopyAttributes.StampPosition.BOTTOM_LEFT.name());
            } else {
                stamptype.remove(CopyAttributes.StampPosition.BOTTOM_LEFT.name());
            }
        }
        if (bottom_right_stampOption != null) {
            if (!bottom_right_stampOption.type.name().equals("NONE")) {
                stamptype.add(CopyAttributes.StampPosition.BOTTOM_RIGHT.name());
            } else {
                stamptype.remove(CopyAttributes.StampPosition.BOTTOM_RIGHT.name());
            }
        }
        if (bottom_center_stampOption != null) {
            if (!bottom_center_stampOption.type.name().equals("NONE")) {
                stamptype.add(CopyAttributes.StampPosition.BOTTOM_CENTER.name());
            } else {
                stamptype.remove(CopyAttributes.StampPosition.BOTTOM_CENTER.name());
            }
        }
        if (stamptype.isEmpty()) {
            stamptype.add("NONE");
        }

        Handler handler = new Handler();
        Runnable r = (Runnable) (new Runnable() {
            public final void run() {
                StampOptionDialogPreference.this.setSummary(String.join(",",stamptype));
            }
        });
        handler.post(r);
    }
}

File: HPWorkpath_v1.6.3/Samples_Kotlin/ExampleAPIServices/source/CopySample/src/com/hp/workpath/sample/copysample/fragments/StampOptionDialogPreference.kt
package com.hp.workpath.sample.copysample.fragments

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.util.AttributeSet
import androidx.preference.DialogPreference
import androidx.preference.PreferenceManager
import com.google.gson.Gson
import com.hp.workpath.api.copier.CopyAttributes
import com.hp.workpath.api.copier.StampOption
import com.hp.workpath.api.copier.StampPolicyType
import com.hp.workpath.api.copier.StampType
import com.hp.workpath.sample.copysample.fragments.CopyConfigureFragment.Companion.PREF_STAMP

class StampOptionDialogPreference(context: Context?, attrs: AttributeSet?) :
    DialogPreference(context, attrs) {
    private var mStampPositionList: List<CopyAttributes.StampPosition>? = null
    private var mStampTypeListMap: MutableMap<CopyAttributes.StampPosition, List<StampType>>? = null
    private var mStampPolicyTypeListMap: MutableMap<CopyAttributes.StampPosition, List<StampPolicyType>>? =
        null
    private var mStampFontListMap: MutableMap<CopyAttributes.StampPosition, List<String>>? = null
    private var mStampTextSizeListMap: MutableMap<CopyAttributes.StampPosition, List<Int>>? = null
    private var mStampTextColorListMap: MutableMap<CopyAttributes.StampPosition, List<String>>? = null
    private var mStampWhiteBackGroundListMap: MutableMap<CopyAttributes.StampPosition, List<Boolean>>? = null
    private var mStampOptionMap: MutableMap<CopyAttributes.StampPosition, StampOption>? = null
    fun getmStampPositionList(): List<CopyAttributes.StampPosition>? {
        return mStampPositionList
    }

    fun setmStampPositionList(mStampPositionList: List<CopyAttributes.StampPosition>?) {
        this.mStampPositionList = mStampPositionList
    }

    fun getmStampTypeListMap(): Map<CopyAttributes.StampPosition, List<StampType>>? {
        return mStampTypeListMap
    }

    fun setmStampTypeListMap(mStampTypeListMap: MutableMap<CopyAttributes.StampPosition, List<StampType>>?) {
        this.mStampTypeListMap = mStampTypeListMap
    }

    fun getmStampPolicyTypeListMap(): Map<CopyAttributes.StampPosition, List<StampPolicyType>>? {
        return mStampPolicyTypeListMap
    }

    fun setmStampPolicyTypeListMap(mStampPolicyTypeListMap: MutableMap<CopyAttributes.StampPosition, List<StampPolicyType>>?) {
        this.mStampPolicyTypeListMap = mStampPolicyTypeListMap
    }

    fun getmStampFontListMap(): Map<CopyAttributes.StampPosition, List<String>>? {
        return mStampFontListMap
    }

    fun setmStampFontListMap(mStampFontListMap: MutableMap<CopyAttributes.StampPosition, List<String>>?) {
        this.mStampFontListMap = mStampFontListMap
    }

    fun getmStampTextSizeListMap(): Map<CopyAttributes.StampPosition, List<Int>>? {
        return mStampTextSizeListMap
    }

    fun setmStampTextSizeListMap(mStampTextSizeListMap: MutableMap<CopyAttributes.StampPosition, List<Int>>?) {
        this.mStampTextSizeListMap = mStampTextSizeListMap
    }

    fun getmStampTextColorListMap(): Map<CopyAttributes.StampPosition, List<String>>? {
        return mStampTextColorListMap
    }

    fun setmStampTextColorListMap(mStampTextColorListMap: MutableMap<CopyAttributes.StampPosition, List<String>>?) {
        this.mStampTextColorListMap = mStampTextColorListMap
    }

    fun getmStampWhiteBackGroundListMap(): MutableMap<CopyAttributes.StampPosition, List<Boolean>>? {
        return mStampWhiteBackGroundListMap
    }

    fun setmStampWhiteBackGroundListMap(mStampWhiteBackGroundListMap: MutableMap<CopyAttributes.StampPosition, List<Boolean>>?) {
        this.mStampWhiteBackGroundListMap = mStampWhiteBackGroundListMap
    }

    fun getmStampOptionMap(): MutableMap<CopyAttributes.StampPosition, StampOption>? {
        return mStampOptionMap
    }

    fun setmStampOptionMap(mStampOptionMap: MutableMap<CopyAttributes.StampPosition, StampOption>?) {
        this.mStampOptionMap = mStampOptionMap
    }

    fun saveStampOptionMap() {
        val sharedPref = PreferenceManager.getDefaultSharedPreferences(
            context
        )
        val editor = sharedPref.edit()
        val gson = Gson()
        for (stampPosition in mStampOptionMap!!.keys) {
            val jsonStampOption = gson.toJson(mStampOptionMap!![stampPosition])
            editor.putString(PREF_STAMP.toString() + stampPosition.name, jsonStampOption)
        }
        editor.apply()
        setStampSummary();
    }


    private fun setStampSummary() {
        val stamptype= mutableListOf<String>()
        val sharedPref = PreferenceManager.getDefaultSharedPreferences(context)
        val gson = Gson()
        val top_left=sharedPref.getString(PREF_STAMP.toString()+CopyAttributes.StampPosition.TOP_LEFT,null)
        val top_right=sharedPref.getString(PREF_STAMP.toString()+CopyAttributes.StampPosition.TOP_RIGHT,null)
        val top_center=sharedPref.getString(PREF_STAMP.toString()+CopyAttributes.StampPosition.TOP_CENTER,null)
        val bottom_left=sharedPref.getString(PREF_STAMP.toString()+CopyAttributes.StampPosition.BOTTOM_LEFT,null)
        val bottom_right=sharedPref.getString(PREF_STAMP.toString()+CopyAttributes.StampPosition.BOTTOM_RIGHT,null)
        val bottom_center=sharedPref.getString(PREF_STAMP.toString()+CopyAttributes.StampPosition.BOTTOM_CENTER,null)
        val top_left_stampOption=gson.fromJson(top_left,StampOption::class.java)
        val top_right_stampOption=gson.fromJson(top_right,StampOption::class.java)
        val top_center_stampOption=gson.fromJson(top_center,StampOption::class.java)
        val bottom_left_stampOption=gson.fromJson(bottom_left,StampOption::class.java)
        val bottom_right_stampOption=gson.fromJson(bottom_right,StampOption::class.java)
        val bottom_center_stampOption=gson.fromJson(bottom_center,StampOption::class.java)

        if(top_left_stampOption!=null){
            if(top_left_stampOption.type!!.name!="NONE"){
                stamptype.add(CopyAttributes.StampPosition.TOP_LEFT.name)
            }else{
                stamptype.remove(CopyAttributes.StampPosition.TOP_LEFT.name)
            }
        }
        if(top_right_stampOption!=null){
            if(top_right_stampOption.type!!.name!="NONE"){
                stamptype.add(CopyAttributes.StampPosition.TOP_RIGHT.name)
            }else{
                stamptype.remove(CopyAttributes.StampPosition.TOP_RIGHT.name)
            }
        }
        if(top_center_stampOption!=null) {
            if (top_center_stampOption.type!!.name != "NONE") {
                stamptype.add(CopyAttributes.StampPosition.TOP_CENTER.name)
            }else{
                stamptype.remove(CopyAttributes.StampPosition.TOP_CENTER.name)
            }
        }
        if(bottom_left_stampOption!=null) {
            if (bottom_left_stampOption.type!!.name != "NONE") {
                stamptype.add(CopyAttributes.StampPosition.BOTTOM_LEFT.name)
            }else{
                stamptype.remove(CopyAttributes.StampPosition.BOTTOM_LEFT.name)
            }
        }
        if(bottom_right_stampOption!=null) {
            if (bottom_right_stampOption.type!!.name != "NONE") {
                stamptype.add(CopyAttributes.StampPosition.BOTTOM_RIGHT.name)
            }else{
                stamptype.remove(CopyAttributes.StampPosition.BOTTOM_RIGHT.name)
            }
        }
        if(bottom_center_stampOption!=null) {
            if (bottom_center_stampOption.type!!.name != "NONE") {
                stamptype.add(CopyAttributes.StampPosition.BOTTOM_CENTER.name)
            }else{
                stamptype.remove(CopyAttributes.StampPosition.BOTTOM_CENTER.name)
            }
        }
        if(stamptype.isEmpty()){
            stamptype.add("NONE")
        }
        val handler = Handler(Looper.getMainLooper())
        val r = Runnable { summary = stamptype.joinToString(",") }
        handler.post(r)
    }
}